paddle/fluid/operators/split_selected_var_op.cc

```diff
@@ -0,0 +1,153 @@
+      auto &to_copy = x->Get<framework::LoDTensor>();
+      auto *out = selected_out->GetMutable<framework::LoDTensor>();
+      framework::TensorCopy(to_copy, to_copy.place(), dev_ctx, out);
+      out->set_lod(to_copy.lod());
+    } else if (x->IsType<framework::SelectedRows>()) {
+      auto &to_copy = x->Get<framework::SelectedRows>();
+      auto *out = selected_out->GetMutable<framework::SelectedRows>();
+      out->set_rows(to_copy.rows());
+      out->set_height(to_copy.height());
+      framework::TensorCopy(to_copy.value(), to_copy.place(), dev_ctx,
+                            out->mutable_value());
+    } else if (x->IsType<framework::LoDTensorArray>()) {
+      auto &to_copy = x->Get<framework::LoDTensorArray>();
+      auto *out = selected_out->GetMutable<framework::LoDTensorArray>();
+      out->resize(to_copy.size());
+      for (size_t i = 0; i < to_copy.size(); ++i) {
+        framework::LoDTensor &out_tensor = (*out)[i];
+        framework::TensorCopy(to_copy[i], to_copy[i].place(), dev_ctx,
+                              &out_tensor);
+        out_tensor.set_lod(to_copy[i].lod());
+      }
+    } else {
+      PADDLE_THROW("Type %s is not supported in MergeSelectedVarOp",
+                   framework::ToTypeName(x->Type()));
+    }
+  }
+
+  // Returns which branch to output
+  int GetOutputBranch(const framework::LoDTensor &mask,
+                      const platform::DeviceContext &dev_ctx) const {
+    PADDLE_ENFORCE_EQ(mask.numel(), 1,
+                      "Mask in SplitSelectedVarOp must have numel 1.");
+    if (platform::is_cpu_place(mask.place())) {
+      return mask.data<int>()[0];
+    }
+    // when platform::is_gpu_place(mask.place()) is ture
+    std::unique_ptr<framework::LoDTensor> cpu_mask{new framework::LoDTensor()};
+#ifdef PADDLE_WITH_CUDA
+    framework::TensorCopy(mask, platform::CPUPlace(), dev_ctx, cpu_mask.get());
+#else
+    PADDLE_THROW(
+        "This version of PaddlePaddle doen NOT support GPU but got GPU tensor "
+        "Mask in SplitSelectedVarOp. Please compile WITH_GPU option");
+#endif
+    return cpu_mask->data<int>()[0];
+  }
+};
+
+class SplitSelectedVarOpProtoMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  void Make() override {
+    AddInput("X", "The input LoDTensor or LoDTensorArray or SelectedRows.");
+    AddInput("Mask", "Tensor with numel 1 specifying which branch to output");
+    AddOutput("Out",
+              "The output can contains multiple variables. The output of "
+              "selected branch will be same as input. We do nothing for "
+              "variables in other branch")
+        .AsDuplicable();
+    // TODO(huihuangzheng): decide whether to add support for lod level
+    // Because this op is blocking whole control flow. I am implementing MVP
+    // (minimal viable product) here.
+    AddComment(R"DOC(
+Split input variable into one output branch. The mask is an integer tensor to
+specify which output branch should copy the input. 
```

Comment:
As discussed offline, we should add some unittests to check whether it is necessary to set the other non-selected var to be zero later.

Reply:
Yes, I will do that in follow-up PRs and you will be one of reviewers :-)